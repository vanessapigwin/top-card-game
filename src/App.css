#root {
  display: flex;
  flex-direction: column;
  height: 100vh;
  max-width: 1080px;
  margin: 0 auto;
  padding: 2rem;
  gap: 2rem;
  text-align: center;
}

.container {
  padding: 2em;
  background-color: rgba(255, 255, 255, 0.87);
  border: solid 1px #f2f2f2;
  border-radius: 1em;
}

.container.title {
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 1.5em;
}

.scorecard {
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: end;
}

.container.card-area {
  flex: 1;
  display: grid;
  grid-template-columns: repeat(auto-fit, 200px);
  grid-template-rows: repeat(auto-fit, 200px);
  gap: 1.5em;
  justify-content: center;
  align-content: center;
}

@media screen and (max-width: 900px) {
  .container.title {
    flex-direction: column;
  }
}
