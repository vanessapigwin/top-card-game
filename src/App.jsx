import { useEffect, useState } from "react";
import "./App.css";
import Card from "./components/Card";

function App() {
  const [highScore, setHighScore] = useState(0);
  const [currentScore, setCurrentScore] = useState(0);
  const [selected, setSelected] = useState([]);
  const [imgList, setImgList] = useState([]);
  const url = "https://api.thecatapi.com/v1/images/search?limit=10";

  function dealCards(selectedCards, apiCards) {
    if (selectedCards.length === 0) return apiCards;

    const cards = apiCards;
    const randomSelection = Math.floor(Math.random() * selectedCards.length);
    const replaceIndex = Math.floor(Math.random() * 8);
    cards[replaceIndex] = selectedCards[randomSelection];
    return cards;
  }

  function selectCard(e) {
    const link = e.target.src;

    if (selected.includes(link)) {
      setHighScore(currentScore);
      setCurrentScore(0);
      setSelected([]);
      alert("Meow! Game Over!");
    } else {
      const newSelected = [...selected, link];
      setSelected(newSelected);
      setCurrentScore(currentScore + 1);
    }
  }

  useEffect(() => {
    let ignore = false;

    if (!ignore) {
      fetch(url)
        .then((response) => {
          return response.json();
        })
        .then((data) => {
          const urlList = data.map((d) => d.url);
          urlList.length = 8;
          const newCards = dealCards(selected, urlList);
          setImgList(newCards);
        });
    }

    return () => {
      ignore = true;
    };
  }, [selected]);

  return (
    <>
      <div className="container title">
        <h1>Cat Memo Game</h1>
        <div className="scorecard">
          <span>Current Score: {currentScore}</span>
          <span>High Score: {highScore}</span>
        </div>
      </div>

      <div className="container card-area">
        {imgList.map((img) => (
          <Card
            imgSrc={img}
            handleClick={selectCard}
            key={crypto.randomUUID()}
          ></Card>
        ))}
      </div>
    </>
  );
}

export default App;
