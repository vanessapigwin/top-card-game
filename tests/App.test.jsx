/*global global*/
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../src/App";

const newfakePromise = (filename) => {
  return new Promise((resolve) =>
    resolve({
      ok: true,
      json: () => Promise.resolve(Array(8).fill({ url: `${filename}.jpg` })),
    }),
  );
};

describe("App component initial page load", () => {
  beforeEach(() => {
    // render the page
    render(<App />);
    // spying on fetch sample
    vi.spyOn(global, "fetch").mockImplementation(() =>
      newfakePromise("kitty1"),
    );
  });

  afterEach(() => vi.clearAllMocks());

  it("Render correct title", () => {
    expect(screen.getByRole("heading").textContent).toMatch(/Plz Kitty/i);
  });

  it("Default render", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it("Let there be 8 img elements", async () => {
    await waitFor(() => {
      const imagesElement = screen.getAllByRole("img");
      expect(imagesElement.length).toBe(8);
    });
  });

  it("let there be actual images for each image element", async () => {
    await waitFor(() => {
      const imagesElement = screen.getAllByRole("img");
      const imgLinks = imagesElement.map((element) =>
        element.getAttribute("src"),
      );
      expect(imgLinks.length).toBe(8);
      expect(imgLinks).toStrictEqual(Array(8).fill("kitty1.jpg"));
    });
  });
});

describe("Click kitties behavior", () => {
  afterEach(() => vi.clearAllMocks());

  it("Click on image will load new images", async () => {
    const mockFetch = vi
      .spyOn(global, "fetch")
      .mockImplementation(() => newfakePromise("kitty2"));

    const user = userEvent.setup();
    let selected;
    render(<App />);

    expect(mockFetch).toHaveBeenCalledOnce();
    waitFor(() => {
      selected = screen.getByRole("img");
      expect(selected.getAttribute("src")).toBe("kitty2.jpg");

      user.click(selected);
      expect(mockFetch).toHaveBeenCalledTimes(2);
    });
  });
});
