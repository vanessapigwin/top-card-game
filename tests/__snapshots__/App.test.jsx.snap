// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > Call cats on render 1`] = `
<div>
  <div
    class="container title"
  >
    <h1>
      Plz Kitty Memo Game
    </h1>
    <div
      class="scorecard"
    >
      <span>
        Current Score: 
        0
      </span>
      <span>
        High Score: 
        0
      </span>
    </div>
  </div>
  <div
    class="container card-area"
  />
</div>
`;

exports[`App component > Default render 1`] = `
<div>
  <div
    class="container title"
  >
    <h1>
      Plz Kitty Memo Game
    </h1>
    <div
      class="scorecard"
    >
      <span>
        Current Score: 
        0
      </span>
      <span>
        High Score: 
        0
      </span>
    </div>
  </div>
  <div
    class="container card-area"
  />
</div>
`;

exports[`App component initial page load > Default render 1`] = `
<div>
  <div
    class="container title"
  >
    <h1>
      Plz Kitty Memo Game
    </h1>
    <div
      class="scorecard"
    >
      <span>
        Current Score: 
        0
      </span>
      <span>
        High Score: 
        0
      </span>
    </div>
  </div>
  <div
    class="container card-area"
  />
</div>
`;
